Add tikv write stall blog (#747)
pingcap/blog: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md

---

**Author:** Mike Barlow (Solutions Architect at PingCAP)

**Editors:** Jose Espinoza, Tina Yang, Tom Dewan

![How to Troubleshoot RocksDB Write Stalls in TiKV](media/troubleshoot-rocksdb-write-stalls-in-tikv.png)

[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.

We will also discuss how to resolve this issue in [TiKV](https://tikv.org/), a highly scalable, low latency, and easy to use key-value database that [uses RocksDB as its storage engine](https://pingcap.com/blog/rocksdb-in-tikv).

When RocksDB can't flush and compact data promptly, it uses a feature called "stalls" to try and slow the amount of data coming into the engine.  Write stalls include pausing all writes and limiting the number of writes.

> From [Github](https://github.com/facebook/rocksdb/wiki/Write-Stalls):

Reviewer: ```suggestion
From [GitHub](https://github.com/facebook/rocksdb/wiki/Write-Stalls):
```

>
> * Increase space amplification, which could lead to running out of disk space;
> * Increase read amplification, significantly degrading read performance.
>
> The idea is to slow down incoming writes to the speed that the database can handle. However, sometimes the database can be too sensitive to a temporary write burst, or underestimate what the hardware can handle, so that you may get unexpected slowness or query timeouts.

TiKV has two instances of RocksDB (RaftDB and KVdb). RaftDB has only one column family. KVDB has three column families (default, write, and lock).

## Write stall user experiences 

Because TiKV depends on RocksDB, when the application or process does a large number of writes, TiDB response time (latency) can degrade significantly. This latency seems to affect all writes and not just writes to a specific table.

There are a couple of Grafana graphs that can help identify whether RocksDB's write stall is the cause and needs additional investigation. 

> We can use the following Grafana charts located in the section 

Reviewer: ```suggestion
We can use the following Grafana charts located in the section:
```

`[Cluster name] / TiKV-Details / [RocksDB - raft | RocksDB -kv]`  

In the chart below, we see that write stalls spiked significantly at around 11:30. Before 11:30 and after 12:15 write stalls did not occur. A healthy system should not have write stalls. 

![Write stall duration](media/write-stall-duration.png)
<div class="caption-center">Write stall duration</div>

With this knowledge, we can dig deeper into write stalls.

RocksDB will trigger write stalls if there are too many:

> * [Memtables](#bookmark=id.hlmn1aylq4qf)
> * [Level 0 SST (Sorted String Table) files](#bookmark=id.cye5mbth2h5x) 
> * [Pending compaction bytes](#bookmark=id.j7eeqy4rw57t)

Reviewer 2: ```suggestion
* [Memtables](#too-many-memtables)
* [Level 0 SST (Sorted String Table) files](#too-many-level-0-sst-files) 
* [Pending compaction bytes](#too-many-pending-compaction-bytes)
```

There are background tasks (jobs) that are specifically responsible for flushing memtables to disk and compacting SST files. By default `max-background-jobs` is set to 8 or CPU cores to -1, whichever is smaller. 

There can be many memtables. Remember TiDB has a minimum of one memtable for each column family (default, write, lock, and raft). As mentioned before, when the memtable reaches the `write-buffer-size` limit it becomes read-only and is flagged to be flushed to disk as an SST file. There are a limited number of tasks that flush memtables to disk, which can be identified with `max-background-flushes` limit. The default is 2 or max-background-jobs / 4, whichever is bigger. 

> **NOTE:**
>

> >Prior to TiKV v4.0.9, `max-background-flushes` is not available, and the number of background flush tasks is set to `max-background-jobs` / 4. 

Reviewer: ```suggestion
> Prior to TiKV v4.0.9, `max-background-flushes` is not available, and the number of background flush tasks is set to `max-background-jobs` / 4. 
```

If the number of memtables reaches the `max-write-buffer-number` for a column family, RocksDB stalls all writes for that column family on that instance.

Configuration parameters:

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].write-buffer-size`

     The default 128 MB. If you increase this parameter, the possibility of an OOM also increases.

> * `[rocksdb|raftdb].[defaultcf|writecf|lockcf]max-write-buffer-number` 
> 
>     The default is 5. If you increase this parameter, the possibility of an OOM also increases
> * `[rocksdb|raftdb].max-background-jobs` 
> 
>     For TiDB 5.0, the default is 8.
> * `[rocksdb|raftdb].max-background-flushes` 
> 
>     The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.

Reviewer 2: ```suggestion

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf]max-write-buffer-number` 

    The default is 5. If you increase this parameter, the possibility of an OOM also increases.

* `[rocksdb|raftdb].max-background-jobs` 

    For TiDB 5.0, the default is 8.

* `[rocksdb|raftdb].max-background-flushes` 

    The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.
```

> **NOTE:**
>
> It's normal for level 0 and level 1 SST files to not be compressed, which is in contrast to other levels that do compress SST files.

By default, when the count of level 0 SST files reaches 20, RocksDB will slow writes for the entire database. If the count of level 0 SST files reaches 36, RocksDB will stall writes for the entire database.

Compaction is triggered when the number of level 0 SST files for a column family reaches 4. However, compaction may not occur. Parallel compactions on different column families can occur simultaneously.

Configuration parameters:

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-file-num-compaction-trigger` 

>     The default is 4.

Reviewer 2: ```suggestion
    The default is 4.
    
```

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-slowdown-writes-trigger`

>     The default is 20.

Reviewer 2: ```suggestion
    The default is 20.
    
```

RocksDB uses level compaction for levels 1 and lower. By default, level 1 has a target size compaction at 512 MB for write and default column families (CFs), and the lock CF has a default of 128 MB. Each lower level has a target size 10 times greater than the previous higher level.  For example, if the level 1 target size is 512 MB, the level 2 target size is 5 GB, the level 3 target Size is 50 GB, and so on.  

When a level exceeds its target size, the excess is known as pending compaction bytes. Pending compaction bytes for levels 1 and lower are aggregated to calculate the total pending compaction bytes. Write stalls are triggered if the total pending compaction bytes exceeds the soft-pending-compaction-bytes-limit or hard-pending-compaction-bytes-limit.

For example, in the diagram below, the total pending compaction is 72 GB. If the soft-pending-compaction-bytes-limit is set to 64 GB (default), write stalls will be triggered.

![Total pending compaction](media/total-pending-compaction.png)
<div class="caption-center">Total pending compaction</div>

Configuration parameters:

* `[rocksdb|raftdb].soft-pending-compaction-bytes-limit` 

>     The default is 64 GB.

Reviewer 2: ```suggestion
    The default is 64 GB.
    
```

---
title: How to Troubleshoot RocksDB Write Stalls in TiKV
author: ['Mike Barlow']
date: 2021-11-12
summary: This troubleshooting guide discusses write performance degradation in TiKV related to the Rocks DB built-in write stall feature.

> tags: ['TiKV']

Reviewer 2: ```suggestion
tags: ['TiKV', 'Tutorial']
```

categories: ['Engineering']
image: /images/blog/troubleshoot-rocksdb-write-stalls-in-tikv.png
---

**Author:** Mike Barlow (Solutions Architect at PingCAP)

> **Editor:** Jose Espinoza, Tina Yang, Tom Dewan

Reviewer 2: ```suggestion
**Editors:** Jose Espinoza, Tina Yang, Tom Dewan
```

There are background tasks (jobs) that are specifically responsible for flushing memtables to disk and compacting SST files. By default `max-background-jobs` is set to 8 or CPU cores to -1, whichever is smaller. 

There can be many memtables. Remember TiDB has a minimum of one memtable for each column family (default, write, lock, and raft). As mentioned before, when the memtable reaches the `write-buffer-size` limit it becomes read-only and is flagged to be flushed to disk as an SST file. There are a limited number of tasks that flush memtables to disk, which can be identified with `max-background-flushes` limit. The default is 2 or max-background-jobs / 4, whichever is bigger. 

> > **NOTE**

Reviewer 2: ```suggestion
> **NOTE:**
```

>
>Prior to TiKV v4.0.9, `max-background-flushes` is not available, and the number of background flush tasks is set to `max-background-jobs` / 4. 

If the number of memtables reaches the `max-write-buffer-number` for a column family, RocksDB stalls all writes for that column family on that instance.

Configuration parameters:

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].write-buffer-size`

     The default 128 MB. If you increase this parameter, the possibility of an OOM also increases.
* `[rocksdb|raftdb].[defaultcf|writecf|lockcf]max-write-buffer-number` 

    The default is 5. If you increase this parameter, the possibility of an OOM also increases
* `[rocksdb|raftdb].max-background-jobs` 

    For TiDB 5.0, the default is 8.
* `[rocksdb|raftdb].max-background-flushes` 

    The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.

Grafana graphs:

![Immutable memtable number](media/immutable-memtable-number.png)

> <div class="caption-center">Immutable memtable number</div> _

Reviewer 2: ```suggestion
<div class="caption-center">Immutable memtable number</div>
```

> **Note:**
>
> It’s normal for level 0 and level 1 SST files to not be compressed, which is in contrast to other levels that do compress SST files.

By default, when the count of level 0 SST files reaches 20, RocksDB will slow writes for the entire database. If the count of level 0 SST files reaches 36, RocksDB will stall writes for the entire database.

Compaction is triggered when the number of level 0 SST files for a column family reaches 4. However, compaction may not occur. Parallel compactions on different column families can occur simultaneously.

Configuration parameters:

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-file-num-compaction-trigger` 

    The default is 4.

> * `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-slowdown-writes-trigger`**` 

Reviewer 2: ?

Author: fixed.

---
title: How to Troubleshoot RocksDB Write Stalls in TiKV
author: ['Mike Barlow']
date: 2021-11-12

> summary: This troubleshooting guide discusses write performance degradation in TiKV related to the Rocks DB built-in write stall feature.

Reviewer 3: ```suggestion
summary: This troubleshooting guide discusses write performance degradation in TiKV related to the RocksDB built-in write stall feature.
```

tags: ['TiKV','Tutorial']
categories: ['Engineering']
image: /images/blog/troubleshoot-rocksdb-write-stalls-in-tikv.png
---

**Author:** Mike Barlow (Solutions Architect at PingCAP)

**Editors:** Jose Espinoza, Tina Yang, Tom Dewan

![How to Troubleshoot RocksDB Write Stalls in TiKV](media/troubleshoot-rocksdb-write-stalls-in-tikv.png)

> [TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.

Author: ```suggestion
[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open-source, mature, and high-performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
```

> [TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.

Reviewer 3: ```suggestion
[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [RocksDB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open-source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
```

> **NOTE:**
>
> It’s normal for level 0 and level 1 SST files to not be compressed, which is in contrast to other levels that do compress SST files.

By default, when the count of level 0 SST files reaches 20, RocksDB will slow writes for the entire database. If the count of level 0 SST files reaches 36, RocksDB will stall writes for the entire database.

Compaction is triggered when the number of level 0 SST files for a column family reaches 4. However, compaction may not occur. Parallel compactions on different column families can occur simultaneously.

Configuration parameters:

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-file-num-compaction-trigger` 

    The default is 4.

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-slowdown-writes-trigger`

    The default is 20.

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-stop-writes-trigger` 

    The default is 36.

Grafana graphs:

![Level 0 SST file number](media/level0-sst-file-number.png)
<div class="caption-center">Level 0 SST file number</div>

### Too many pending compaction bytes

> A Log-Structured Merge-Tree (LSM) consists of multiple levels. Each level can include zero or more SST files. Level 0 is identified as a "higher hevel" and each incrementing level (level 1, level 2, etc.) is identified as a lower level. (See the diagram below.)

Reviewer: ```suggestion
A Log-Structured Merge-Tree (LSM) consists of multiple levels. Each level can include zero or more SST files. Level 0 is identified as a "higher level" and each incrementing level (level 1, level 2, etc.) is identified as a lower level. (See the diagram below.)
```
